### PR TITLE
Correct issue with Ford vehicles where after a human initiated turn they will not straighten up, requiring drive to fight the wheel for 3-5 seconds.

### PR DESCRIPTION
**Pull request by alan-polk**
<!--- ***** Template: Car Bugfix *****

**Description**

If Ford vehicles which are curvature based, if the driver takes over the steering wheel and makes a turn with OP lateral engaged (this happens in experimental mode at red lights and stop signs), the curvature will build up in the EPAS and the car will continue to try and turn after the maneuver is over, requiring the driver to fight the steering wheel until the curvature inside the EPAS completes.  This usually last 3-5 seconds, where if the driver goes hands free again, the car will exit the road in the direction the turn was navigated.

An analysis in PlotJuggler of Ford logic from dash cam routes on F150s, Mavericks, and Mach E's shows that Ford accounts for this by changing the ramp type to "immediate" and setting curvature to zero whenever it detects steering wheel pressed and the steering wheel angle is greater than 60 degrees.  This PR attempts to replicate that logic.

**Verification**

Tested on 3 F150 routes, 1 Edge Route, 1 Maverick Route, 1 Mach E Route.

**Route**

e36b272d5679115f/2024-02-19--12-00-50/1

**Verification**

Tested by 6 different drivers.  We have not found highway speed curves which result in > 60 degree wheel angle, but even if one did occur, unless the wheel is also touched with more than 1nM of force, this logic would not engage.  Ford stock logic is always in hands on mode during such a curve.

### CODE REVIEW

File: selfdrive/car/ford/carcontroller.py [hunk 1]
```diff
@@ -69,13 +69,22 @@ def update(self, CC, CS, now_nanos):
 
       self.apply_curvature_last = apply_curvature
 
+      steeringPressed = CS.out.steeringPressed
+      steeringAngleDeg = CS.out.steeringAngleDeg
+
+      if steeringPressed and abs(steeringAngleDeg) > 60:
+        apply_curvature = 0
```
Comment (incognitojam) [inline selfdrive/car/ford/carcontroller.py:76]: This will trip the panda safety causing messages to be dropped and the system will stop applying steering. Have you checked the `totalTxBlocked` in plotjuggler?

**Comment by incognitojam** [comment]
How does this work? Are you trying to make the system steer to center or does setting the ramp type to immediate do something here? I suspect that your messages are being blocked because you make a sudden change to the applied curvature. If the PSCM stops receiving messages it will stop applying torque but not until a few seconds have passed.

**Comment by coffee-cake-isaac** [comment]
> How does this work? Are you trying to make the system steer to center or does setting the ramp type to immediate do something here? I suspect that your messages are being blocked because you make a sudden change to the applied curvature. If the PSCM stops receiving messages it will stop applying torque but not until a few seconds have passed.

It's much less to do with actual control of the car and moreso what happens when we take control to complete turns that OP cannot. The issue, at least with Q4 Fords (I can't speak to Q3), is that when we complete a 90 degree turn, for instance, the wheel will be held at a high steering angle of the just completed turn for multiple seconds afterwards. We have to manually hold the wheel at center to not run off the road. 

This is a Ford issue, and from what I've talked about with AJ -- it's handled similarly with Ford's stock system. @alan-polk it'd probably help to share your plotjuggler/cabana graphs here.

Basically, the goal for this is to actually *stop* sending messages when the steering angle is greater than 60 degrees and the driver has manually intervened. The intent here is that stopping the lateral messages prevents the PSCM from trying to hold the wheel.

File: selfdrive/car/ford/carcontroller.py (hunk 1 above)
Comment (alan-polk) [reply selfdrive/car/ford/carcontroller.py:76]: I have not checked totalTxBlocked, but we are not receiving any of the warnings on screen "steering temporarily unavailable" that we typically see when we run afoul of panda.

**Comment by alan-polk** [comment]
> How does this work? Are you trying to make the system steer to center or does setting the ramp type to immediate do something here? I suspect that your messages are being blocked because you make a sudden change to the applied curvature. If the PSCM stops receiving messages it will stop applying torque but not until a few seconds have passed.

I'm not trying to make it steer to center, simply resetting the EPAS so it doesn't continue to steer after the turn is complete, and attempting to duplicate the logic (best as I could reverse engineer it) Ford uses to accomplish the same thing.  I'll put together some plots this afternoon and post them. I suspect this is an issue you don't see very often since you drive a stick-shift.  You will typically have to drop the clutch to make a turn affected by this and OP will have already disengaged.

**Comment by incognitojam** [comment]
> I'm not trying to make it steer to center, simply resetting the EPAS so it doesn't continue to steer after the turn is complete

Does setting `LatCtl_D_Rq` to 0 not achieve the same thing, and is clearer?

> and attempting to duplicate the logic (best as I could reverse engineer it) Ford uses to accomplish the same thing

Do you have logs to share from the stock system when the driver is overriding?

File: selfdrive/car/ford/carcontroller.py (hunk 1 above)
Comment (sshane) [reply selfdrive/car/ford/carcontroller.py:76]: This just likely prevents wind up while steering at high curvatures. The issue isn't Ford specific, it's a problem with any low-rate angle control car, so I would like to avoid make-specific logic.

Comment (alan-polk) [reply selfdrive/car/ford/carcontroller.py:76]: I tested Cameron's suggestion of disabling lateral during human turns and achieved the same result of eliminating the problem.  I have not experienced this issues in other cars (I have only driven my F150 and a Santa Fe on OP).  But it clearly isn't as severe on other cars or you would not be making a suggestion to close this PR, because it an unacceptable event.  If you wish for me to move this code to a place where it affects all makes and models I am okay with that. If you wish for me to update using lateral active to achieve the result instead of ramp_type and curvature, I'm okay with that too.  I will make a video tonight of how this behaves on stock OP and post it here

**Comment by FlyMyPG** [comment]
I get this on my '21 Nissan LEAF, so it may not be Ford-specific. I have some things to check in 0.9.6 and will update this comment right after the release.

File: selfdrive/car/ford/carcontroller.py (hunk 1 above)
Comment (alan-polk) [reply selfdrive/car/ford/carcontroller.py:76]: Here is a route exhibiting the issues with stock OP logic:
24574459dd7fb3e0/2024-02-20--19-24-10/0

Here is a video showing the wheel action for 3 of the turns:
https://youtu.be/6yM6Kt6ibwA

Here is a route with the Cameron's suggest logic to disable lateral during the turn:
24574459dd7fb3e0/2024-02-20--19-44-38/0

Here is the video associated with some turns on that route:
https://youtu.be/TcXg6q8q42o

Since we have already moved away from the idea of using ramp_type, I have not added extraneous data related to Ford's stock logic here.

If this is an issue that affects multiple vehicles, I can cancel this PR and create a new one that makes the logic universal.  If after looking at the video you believe it to be a Ford issue, then I will update the PR with the latest logic to cancel lateral instead of use ramp_type.  I'm good with either one, but the issue addressed by this PR can be ignored.

**Comment by FlyMyPG** [comment]
> I get this behavior on my '21 Nissan LEAF, so it may not be Ford-specific. I have some things to check in 0.9.6 and will update this comment right after the release.

Still seeing the same behavior with 0.9.6.

**Comment by sshane** [comment]
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.

**Comment by Rekkur150** [comment]
@alan-polk out of curiosity has this issue gone anywhere?

**Comment by alan-polk** [comment]
> @alan-polk out of curiosity has this issue gone anywhere?

I believe if we want Comma to consider to consider it, I have to resubmit the PR into the openDBC repository since they moved the code over there.  I've been recently working on disabled lateral instead of the RAMP_TYPE solution since Shane appeard more amenable to that slution.

**Comment by alan-polk** [comment]
> @alan-polk I am interested getting this issue fixed so you need any help let me know

what vehicle are you driving?

**Comment by Rekkur150** [comment]
> > @alan-polk I am interested getting this issue fixed so you need any help let me know
> 
> what vehicle are you driving?

Me Chevy bolt (I know not applicable) my friend here has me looking into this, so I am gunna volunteer him. @wiseguy56 he has a 2021 ford explorer

**Comment by alan-polk** [comment]
For the Explorer, this logic is incorporated into bluepilot.

https://github.com/BluePilotDev/bluepilot

installer URL: bit.ly/bp-stable

**Comment by alan-polk** [comment]
> > > @alan-polk I am interested getting this issue fixed so you need any help let me know
> > 
> > 
> > what vehicle are you driving?
> 
> Me Chevy bolt (I know not applicable) my friend here has me looking into this, so I am gunna volunteer him. @wiseguy56 he has a 2021 ford explorer

For the Explorer, this logic is incorporated into bluepilot.

https://github.com/BluePilotDev/bluepilot

installer URL: bit.ly/bp-stable

**Comment by wiseguy56** [comment]
The Bluepilot logic works well
